pkg/controller/storagenode/storagenode.go

```diff
@@ -0,0 +1,393 @@
+
+		if !isBeingDeleted && len(kvdbPodList.Items) == 0 {
+			pod, err := c.createKVDBPod(cluster, storageNode)
+			if err != nil {
+				return err
+			}
+
+			c.log(storageNode).Infof("creating kvdb pod: %s/%s", pod.Namespace, pod.Name)
+			err = c.client.Create(context.TODO(), pod)
+			if err != nil {
+				return err
+			}
+		}
+	} else { // delete pods if present
+		for _, p := range kvdbPodList.Items {
+			c.log(storageNode).Debugf("deleting kvdb pod: %s/%s", p.Namespace, p.Name)
+			err = c.client.Delete(context.TODO(), &p)
+			if err != nil {
+				c.log(storageNode).Warnf("failed to delete pod: %s/%s due to: %v", p.Namespace, p.Name, err)
+			}
+		}
+	}
+	return nil
+}
+func (c *Controller) syncStorage(
+	cluster *corev1alpha1.StorageCluster,
+	storageNode *corev1alpha1.StorageNode,
+) error {
+	// sync the storage labels on pods
+	portworxPodList := &v1.PodList{}
+	pxLabels := c.Driver.GetSelectorLabels()
+	err := c.client.List(
+		context.TODO(),
+		portworxPodList,
+		&client.ListOptions{
+			Namespace:     storageNode.Namespace,
+			LabelSelector: labels.SelectorFromSet(pxLabels),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get list of portworx pods. %v", err)
+	}
+
+	for _, pod := range portworxPodList.Items {
+		podCopy := pod.DeepCopy()
+		controllerRef := metav1.GetControllerOf(podCopy)
+		if controllerRef != nil && controllerRef.UID == cluster.UID &&
+			len(pod.Spec.NodeName) != 0 && storageNode.Name == podCopy.Spec.NodeName {
+			updateNeeded := false
+			value, present := podCopy.GetLabels()[constants.LabelKeyStoragePod]
+			if canNodeServeStorage(storageNode) { // node has storage
+				if value != constants.LabelValueTrue {
+					if podCopy.Labels == nil {
+						podCopy.Labels = make(map[string]string)
+					}
+					podCopy.Labels[constants.LabelKeyStoragePod] = constants.LabelValueTrue
+					updateNeeded = true
+				}
+			} else if present {
+				c.log(storageNode).Debugf("removing storage label from pod: %s/%s",
+					podCopy.Namespace, pod.Name)
+				delete(podCopy.Labels, constants.LabelKeyStoragePod)
+				updateNeeded = true
+			}
+
+			if updateNeeded {
+				if err := c.client.Update(context.TODO(), podCopy); err != nil {
+					return err
+				}
+			}
+			break // found pod we were looking for, no need to check other pods
+		}
+	}
+	return nil
+}
+
+func (c *Controller) createKVDBPod(
+	cluster *corev1alpha1.StorageCluster,
+	storageNode *corev1alpha1.StorageNode,
+) (*v1.Pod, error) {
+	podSpec, err := c.Driver.GetKVDBPodSpec(cluster, storageNode.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kvdb pod template: %v", err)
+	}
+
+	k8s.AddOrUpdateStoragePodTolerations(&podSpec)
+	podSpec.NodeName = storageNode.Name
+
+	trueVar := true
+	newPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-kvdb-", c.Driver.String()),
+			Namespace:    storageNode.Namespace,
+			Labels:       c.kvdbPodLabels(cluster),
+			OwnerReferences: []metav1.OwnerReference{
+				{
```

Comment:
nit: We could use this - `metav1.NewControllerRef(storageNode, corev1alpha1.SchemeGroupVersion.WithKind("StorageNode"))`, instead of manually creating the owner ref.